### ISSUE TITLE
Unresolved dependencies in Dagger 2.41

### ISSUE THREAD

**Issue by Lingviston** [google/dagger#3357]
Yet another issue, which might be related to the annotation processing fix in 2.41.

I heavily use component dependencies in the project. Here's how my setup looks like:
`:app`->`:app-core`
`:app`->`:llibrary`->`:app-core`

In all 3 modules the `javax.inject` dependency is present and dagger compiler is added into the kapt configuration.

`:app-core` includes the `WorkManager` dependency, defines dagger component and exposes the `DelegatingWorkerFactory` class of WorkManager via its API. It also provides a `Foo` dependency. Here's how it looks like:
```
interface Foo { ... }

@Module
internal interface CoreAppModule {
    
    @Provides
    @Singleton
    fun provideDelegatingWorkerFactory(): DelegatingWorkerFactory = DelegatingWorkerFactory()

    @Provides
    fun provideFoo(): Foo = Foo()
}

interface CoreAppApi {

    fun delegatingWorkerFactory(): DelegatingWorkerFactory

    fun foo(): Foo
}

@Component(modules = [CoreAppModule::class])
interface CoreAppComponent : CoreAppApi
```

Module `:library` has its own component, which depends on the `CoreAppApi`, but only consumes the `Foo` from it. Here's how it looks like:

```
@Component(dependencis = [CoreAppApi::class])
internal interface LibraryComponent {
    
    fun foo():Foo
}
```

On attempt to generate the component class for `LibraryComponent` Dagger fails, saying that:
```
error: ComponentProcessingStep was unable to process 'LibraryComponent' because 'androidx.work.DelegatingWorkerFactory' could not be resolved.
```

I oversimplified the setup for the post, but hope it's still understandable. In the real project the `CoreAppApi` exposes much more classes and there're much more `:library` components, which consume only parts of it. All of them fail to compile because of `DelegatingWorkerFactory` not being present in their classpath. I assume that the fix would be to add `WorkManager` dependency into all of the `:library` modules, but this is what I don't want to do. Perhaps the right choice would be to wrap `DelegatingWorkerFactory` by a custom public interface in the `:core-app`.
But before doing any of that I'd like to know if this is really considered to be an issue. I'd expect Dagger to not complain on `LibraryComponent` not being able to resolve types, which it does not consume.

**Comment by Lingviston** [comment]
@danysantiago the reason of the issue is clear for me. What is not clear is why it started being reported only after updating the Dagger to 2.41. I would expect Kotlin/Java compiler to complain on this long ago, not the Dagger processor.

**Comment by Lingviston** [comment]
And IMO, the processor should've only started complaining about the unresolved **annotation** classes, not **all** classes.

**Comment by bcorso** [comment]
> What is not clear is why it started being reported only after updating the Dagger to 2.41. I would expect Kotlin/Java compiler to complain on this long ago, not the Dagger processor.

The Kotlin/Java compiler are just compiling your sources, so they will not fail in this case because your sources do not directly reference `DelegatingWorkerFactory`. The `DelegatingWorkerFactory` is referenced by a transitive dependency `ComponentLibrary -> CoreAppModule -> DelegatingWorkerFactory `.

The reason the Dagger processor is now failing in this case is because we've recently become more strict about which types we need to validate to ensure correctness (see related discussion on #3262), which will often require resolving transitive dependencies -- in this case `DelegatingWorkerFactory` because it's the return type of an `@Provides` method..

**Comment by Lingviston** [comment]
Thanks @bcorso! It all makes sense now.
